### PR TITLE
[FLINK-10748] Decouple RestServerEndpoint from Dispatcher

### PR DESCRIPTION
**Pull request by tillrohrmann**
## What is the purpose of the change

This commits decouples the RestServerEndpoint from the Dispatcher by converting the
RedirectHandler into LeaderRetrievalHandler which only retrieves the gateway of the
current leader instead of redirecting the client. This is possible since the serving
RestServerEndpoint no longer needs to be colocated with the Dispatcher and JobMaster
after all RPC requests are serializable (e.g. no direct access on the ExecutionGraph
is needed).

With this change every RestServerEndpoint will automatically proxy to the current leader
without needing to redirect the requesting client.

## Brief change log

- Remove redirection logic from `RedirectHandler`
- Rename `RedirectHandler` into `LeaderRetrievalHandler`
- Remove `restAddressFuture` from all REST handlers except for the `JarListHandler`

## Verifying this change

- Tested manually
- Adapted former `RedirectHandlerTest` which is now called `LeaderRetrievalHandlerTest`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/util/HandlerRedirectUtils.java [hunk 1]
```diff
@@ -52,25 +46,6 @@
...
-					return Optional.of(remoteRestAddress);
-				}
-			});
-	}
-
-	public static HttpResponse getRedirectResponse(String redirectAddress, String path) {
-		return getRedirectResponse(redirectAddress, path, HttpResponseStatus.TEMPORARY_REDIRECT);
-	}
-
 	public static HttpResponse getRedirectResponse(String redirectAddress, String path, HttpResponseStatus code) {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/util/HandlerRedirectUtils.java]: I supposed the remaining ones are used for the SSL redirection?

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/LeaderRetrievalHandler.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+		HttpRequest request = routedRequest.getRequest();
+
+		OptionalConsumer<? extends T> optLeaderConsumer = OptionalConsumer.of(leaderRetriever.getNow());
+
+		optLeaderConsumer.ifPresent(
+			gateway -> {
+				try {
+					respondAsLeader(channelHandlerContext, routedRequest, gateway);
+				} catch (Exception e) {
+					logger.error("Error while responding as leader.", e);
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/LeaderRetrievalHandler.java]: this isn't _quite_ correct is it.

File: flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/LeaderRetrievalHandlerTest.java [hunk 3]
```diff
@@ -101,31 +87,14 @@ public void testRedirectHandler() throws Exception {
...
-			// 3. with leader gateway which is not the one of this REST endpoints --> Redirection required
-			httpClient.sendGetRequest(restPath, FutureUtils.toFiniteDuration(timeout));
-
-			response = httpClient.getNextResponse(FutureUtils.toFiniteDuration(timeout));
-
-			Assert.assertEquals(HttpResponseStatus.TEMPORARY_REDIRECT, response.getStatus());
-			Assert.assertEquals(expectedRedirection, response.getLocation());
-
-			// 4. with local REST endpoint
+			// 2. with leader
```
Comment (zentol) [inline flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/LeaderRetrievalHandlerTest.java]: help me out here, why is the leader now suddenly available?

Comment (tillrohrmann) [reply flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/LeaderRetrievalHandlerTest.java]: with the removal of the redirection logic, there are only two cases:

1. Leader could not be retrieved yet
2. Leader has been retrieved

These two cases are tested.

Comment (zentol) [reply flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/LeaderRetrievalHandlerTest.java]: oh i get that, but this test only executes
```
httpClient.sendGetRequest(restPath, FutureUtils.toFiniteDuration(timeout));

HttpTestClient.SimpleHttpResponse response = httpClient.getNextResponse(FutureUtils.toFiniteDuration(timeout));
```
twice, with each getting a different response. What causes the response to be different? I don't see any modifications being made to the state of the rest server / futures between the 2 requests. This isn't just a timing thing, is it?

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/LeaderRetrievalHandler.java (hunk 2 above)
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/LeaderRetrievalHandler.java]: I'd prefer that. It's a bit....well we just aren't _really_ the leader (in a way this concept no longer exists on the REST handler level), so it seems a bid misleading.

Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/LeaderRetrievalHandler.java]: Alright, will change it.

File: flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/LeaderRetrievalHandlerTest.java (hunk 3 above)
Comment (tillrohrmann) [reply flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/LeaderRetrievalHandlerTest.java]: I actually removed the mocking completely. This should make things a bit clearer.

**Comment by tillrohrmann** [comment]
Thanks for the review @zentol. Merging this PR.

**Pull request merged**